docs/developer-docs/docs/operate-a-node/buenavista-testnet/upgrade/v0.4.0.md

```diff
@@ -4,4 +4,102 @@ sidebar_position: 1
+
+(?) The network upgrade will take place on **August 1, 2024**. The Warden team is going to take the snapshot, export the state, and revert it back if needed.
+
+If you have any outstanding questions, [join our Discord](https://discord.com/invite/warden).
+
+:::warning
+The upgrade procedure carries heightened risks of double-signing and being slashed. To handle these risks and securely address any issues that may arise during the upgrade, follow our tips listed in [Security and troubleshooting](#security--troubleshooting).
+:::
+
+## 1. Verify & back up
+
+Before starting your validator and signing, verify your software version and the genesis file hash.
+
+We also recommend to take a **full data snapshot** prior to the upgrade.
+
+The method depends on your infrastructure, but generally you can achieve this by backing up the **wardend** directory. If you're [using Cosmovisor](#option-2-use-cosmovisor) to upgrade, it'll back up your data by default.
+
+:::warning
+(?) After stopping the `wardend` process, it's critically important to back up the `.wardend/data/priv_validator_state.json` file.
+:::
+
+**Note**: The **priv_validator_state.json**  file is updated every block as your validator participates in consensus rounds. In case the upgrade fails and you have to restart the previous chain, you'll need this file to avoid double signing. See [Handling node restarts](#handling-node-restarts).
+
+## 2. Run Connect sidecar
+
+(?) Now you need to run the Skip:Connect binary on the same machine as `wardend`.
+
+Follow this guide: [Operate Connect](/operate-a-node/operate-connect).
+
+## 3. Upgrade
+
+### Option 1: Upgrade manually
+
+1. (?) Download the [v0.4.0 wardend binary](https://github.com/warden-protocol/wardenprotocol/releases/tag/v0.4.0) or build it from the source code. See the installation instructions in the [Join Buenavista guide](/operate-a-node/buenavista-testnet/join-buenavista#1-install).
+
+2. Run Warden v4.0.0 until the upgrade height. The node will panic:
+
+    (?)
+    
+    ```shell
+    ERR UPGRADE "v0.4.0" NEEDED at height: 12255555: upgrade to v0.4.0 and applying upgrade "v0.4.0" at     height: 12255555
+    ```
+3. Stop the node and switch the binary to **Warden v4.0.1**. Restart the node:  
+
+    ```
+    wardend start
+    ```
+
+4. The chain will continue producing blocks when validators with a total sum voting power > 2/3 complete their node upgrades. It may take from several minutes to a few hours.
+
+### Option 2: Use Cosmovisor
+
+To upgrade using [Cosmovisor](https://pkg.go.dev/cosmossdk.io/tools/cosmovisor), take these steps:
+
+1. Build Warden v0.4.0. and move the `wardend` v0.4.0 binary to `$WARDEN_HOME/cosmovisor/upgrades/v0.4.0/bin`:
+    
+    ```shell
+    mkdir -p  $WARDEN_HOME/cosmovisor/upgrades/v0.4.0/bin
+    cp $(which wardend) $WARDEN_HOME/cosmovisor/upgrades/v0.4.0/bin
+    ```
+
+    **Note**: Alternatively, enable auto-download in Cosmovisor, and it'll automatically download the binary.
```

Comment:
Should we remove this maybe?

`**Note**: Alternatively, enable auto-download in Cosmovisor, and it'll automatically download the binary.`

Maybe its ok on testnets, but for mainnets it is not considered a secure practice to automatically download binaries..

Reply:
No, I'm really arguing against that perspective, we ran 1 year without any issues the automatic update, works flawlessly when the binaries are listed in gov proposal.